File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        
        int n = s.size(), i = 0, j = 0, ans = 0;
        
        unordered_set<char>charSet;
        
        while(j < n){
            
            if(charSet.find(s[j]) == charSet.end()){
                
                charSet.insert(s[j++]);
                
                int size = charSet.size();
                
                ans = max(ans, size);
                
            }else{
                
                while(charSet.find(s[j]) != charSet.end()){
                    
                    charSet.erase(s[i++]);
                    
                }
                
            }
        }
        
        return ans;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        
        int i = 0, j = 0, n = s.length();
        
        int res = 0;
        
        Set<Character>st = new HashSet<>();
        
        while(j < n){
            
            while(i < j && st.contains(s.charAt(j))){
                
                st.remove(s.charAt(i));
                
                ++i;
                
            }
            
            st.add(s.charAt(j));
            
            res = Math.max(res, st.size());
            
            ++j;
            
        }
        
        return res;
        
    }
}